drivers/gpio/gpio_silabs.c

```diff
@@ -0,0 +1,415 @@
+	out = sl_hal_gpio_get_pin_output(&gpio);
+
+	switch (mode) {
+	case SL_GPIO_MODE_WIRED_AND:
+		flags = GPIO_OUTPUT | GPIO_OPEN_DRAIN;
+
+		if (out) {
+			flags |= GPIO_OUTPUT_HIGH;
+		} else {
+			flags |= GPIO_OUTPUT_LOW;
+		}
+		break;
+
+	case SL_GPIO_MODE_WIRED_OR:
+		flags = GPIO_OUTPUT | GPIO_OPEN_SOURCE;
+
+		if (out) {
+			flags |= GPIO_OUTPUT_HIGH;
+		} else {
+			flags |= GPIO_OUTPUT_LOW;
+		}
+		break;
+
+	case SL_GPIO_MODE_PUSH_PULL:
+		flags = GPIO_OUTPUT | GPIO_PUSH_PULL;
+
+		if (out) {
+			flags |= GPIO_OUTPUT_HIGH;
+		} else {
+			flags |= GPIO_OUTPUT_LOW;
+		}
+		break;
+
+	case SL_GPIO_MODE_INPUT_PULL:
+		flags = GPIO_INPUT;
+
+		if (out) {
+			flags |= GPIO_PULL_UP;
+		} else {
+			flags |= GPIO_PULL_DOWN;
+		}
+		break;
+
+	case SL_GPIO_MODE_INPUT:
+		flags = GPIO_INPUT;
+		break;
+
+	case SL_GPIO_MODE_DISABLED:
+		flags = GPIO_DISCONNECTED;
+		break;
+
+	default:
+		break;
+	}
+
+	*out_flags = flags;
+	return 0;
+}
+#endif
+
+static int gpio_silabs_port_get_raw(const struct device *dev, gpio_port_value_t *value)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+
+	*value = sl_hal_gpio_get_port_input(config->gpio_index);
+
+	return 0;
+}
+
+static int gpio_silabs_port_set_masked_raw(const struct device *dev, gpio_port_pins_t mask,
+					   gpio_port_value_t value)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+
+	sl_hal_gpio_set_port_value(config->gpio_index, value, mask);
+
+	return 0;
+}
+
+static int gpio_silabs_port_set_bits_raw(const struct device *dev, gpio_port_pins_t mask)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+
+	sl_hal_gpio_set_port(config->gpio_index, mask);
+
+	return 0;
+}
+
+static int gpio_silabs_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t mask)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+
+	sl_hal_gpio_clear_port(config->gpio_index, mask);
+
+	return 0;
+}
+
+static int gpio_silabs_port_toggle_bits(const struct device *dev, gpio_port_pins_t mask)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+
+	sl_hal_gpio_toggle_port(config->gpio_index, mask);
+
+	return 0;
+}
+
+static int gpio_silabs_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+					       enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	const struct gpio_silabs_port_config *config = dev->config;
+	struct gpio_silabs_port_data *data = dev->data;
+	struct gpio_silabs_common_data *common = data->common_dev->data;
+	sl_gpio_t gpio = {.port = config->gpio_index, .pin = pin};
+	int32_t int_no = SL_GPIO_INTERRUPT_UNAVAILABLE;
+	sl_gpio_interrupt_flag_t flag = SL_GPIO_INTERRUPT_RISING_FALLING_EDGE;
+
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		LOG_ERR("Level interrupt not supported on pin %u", pin);
+		return -ENOTSUP;
+	}
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		ARRAY_FOR_EACH(common->irq_pin_map, i) {
+			if ((data->int_enabled_mask & BIT(i)) && (common->irq_pin_map[i] == pin)) {
```

Comment:
Maybe `irq_pin_map` could be replaced by calls to  `sl_hal_gpio_get_external_interrupt_number()` (this is prefered solution because we prefer a single source of trust).